arangod/RestHandler/RestReplicationHandler.cpp

```diff
@@ -2700,16 +2589,144 @@ ReplicationApplier* RestReplicationHandler::getApplier(bool& global) {
+  }
+  // For leak protection in case of errors the unique_ptr
+  // is not responsible anymore, it has been handed over to the
+  // registry.
+  auto q = query.release();
+  // Make sure to return the query after we are done
+  TRI_DEFER(queryRegistry->close(&_vocbase, id));
+
+  TRI_ASSERT(isLockHeld(id).ok());
+  TRI_ASSERT(isLockHeld(id).get() == false);
+
+  TRI_IF_FAILURE("RestReplicationHandler::destroyBeforeLock") {
+    // This will flag the query as destroyed while getting the
+    // blocking lock
+    cancelBlockingTransaction(id);
+  }
+
+  return q->trx()->begin();
+}
 
-std::unordered_map<std::string, std::shared_ptr<SingleCollectionTransaction>>
-    RestReplicationHandler::_holdReadLockJobs;
+ResultT<bool> RestReplicationHandler::isLockHeld(aql::QueryId id) const {
+  // We try to get a lease of the query.
+  // If this is sucessful we will have the lock,
+  // otherwise we do not have it.
+  // The query is only hold for long during initial locking
+  // there it should return false.
+  // In all other cases it is released quickly.
+  auto queryRegistry = QueryRegistryFeature::QUERY_REGISTRY;
+  TRI_ASSERT(queryRegistry != nullptr);
+  if (queryRegistry == nullptr) {
+    return {TRI_ERROR_SHUTTING_DOWN};
+  }
+  try {
+    auto query = queryRegistry->open(&_vocbase, id);
+    if (query == nullptr) {
+      return ResultT<bool>::error(TRI_ERROR_HTTP_NOT_FOUND, "no hold read lock job found for 'id'");
+    }
+    TRI_DEFER(queryRegistry->close(&_vocbase, id));
+    return true;
+  } catch (...) {
+    // Query exists, but is in use.
+    // So in Locking phase
+    return false;
+  }
+}
+
+ResultT<bool> RestReplicationHandler::cancelBlockingTransaction(aql::QueryId id) const {
+  // This lookup is only required for API compatibility,
+  // otherwise an unconditional destroy() would do.
+  auto res = isLockHeld(id);
+  if (res.ok()) {
+    auto queryRegistry = QueryRegistryFeature::QUERY_REGISTRY;
+    TRI_ASSERT(queryRegistry != nullptr);
+    if (queryRegistry == nullptr) {
+      return {TRI_ERROR_SHUTTING_DOWN};
+    }
+    try {
+      // Code does not matter, read only access, so we can roll back.
+      queryRegistry->destroy(&_vocbase, id, TRI_ERROR_QUERY_KILLED);
+    } catch (...) {
+      // All errors that show up here can only be
+      // triggered if the query is destroyed in between.
+    }
+  }
+  return res;
+}
+
+ResultT<std::string> RestReplicationHandler::computeCollectionChecksum(aql::QueryId id, LogicalCollection* col) const {
+  auto queryRegistry = QueryRegistryFeature::QUERY_REGISTRY;
+  TRI_ASSERT(queryRegistry != nullptr);
+  if (queryRegistry == nullptr) {
+    return {TRI_ERROR_SHUTTING_DOWN};
+  }
+
+  try {
+    auto query = queryRegistry->open(&_vocbase, id);
+    if (query == nullptr) {
+      // Query exists, but is in use.
+      // So in Locking phase
+      return ResultT<std::string>::error(TRI_ERROR_TRANSACTION_INTERNAL, "Read lock not yet acquired!");
```

Comment:
`open` only returns `null` when no query with this id is found; it will throw if the query exists but is already open.
Is this code correct? If yes, then the comment is misleading.

Reply:
Good catch indeed, i run into issue with IsLockHeld during test, but not here and only fixed one of them. (code is closely)
Thanks for checking!!!